Bluetooth: Controller: support FEMs via devicetree

mbolivar-nordic opened the pull request:
The purpose of this pull request is to convert some hardware description in the bluetooth controller subsystem to use devicetree instead of Kconfig. Specifically, radio front end modules (FEMs) are moved to devicetree nodes. The Nordic controller supports this feature. We shouldn't break anything in the openisa controller, but since this controller doesn't have FEM support today, I'm not adding it in.

The first part of the series is prep work:

- update devicetree bindings
- add indirection between existing kconfig options and users in the controller

Then the main patch is `Bluetooth: controller: support FEMs via devicetree`.

We follow that by removing the now-unused Kconfig options.

hubertmis commented:
I like this proposal. It's aligned with the direction I described in #30708.
I also like the ""for production" production" approach :)

anangl reviewed (commented):
LGTM

Nit: the first line of the commit message for the first patch mentions CONFIG_BT_CTLR_GPIO_LNA_PIN instead of CONFIG_BT_CTLR_GPIO_PA_PIN. I actually got confused for a moment when I started reading it. Perhaps it would be good to correct this.

carlescufi commented:
@czeslawmakarski and @ryanjh please take a look

KAGA164 commented:
I like this proposal to unify the configuration of FEM devices. We can then move nRF21540 EK shield configuration from NCS to upstream Zephyr since there will be use case for it. About CONFIG_BT_CTLR_GPIO_PA and CONFIG_BT_CTLR_GPIO_LNA it depends in my opinion if turning off fem in software make sense, for the nRF21540dk if we don't enable fem then radio won't work since there is no bypass for radio signal but I am not familiar with other fem solution so maybe It is worth to keep possibility to turn off fem in software

pdunaj reviewed (changes requested):
nacking to make sure hack is removed.
there seems to be a board that refers to the option being removed:
https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/arm/ubx_bmd345eval_nrf52840/ubx_bmd345eval_nrf52840_defconfig#L34

cvinayak reviewed (commented):
Functional changes to Controller code is nil. Minor explanation requested, otherwise LGTM (not tested myself, hope to do though).

mbolivar-nordic committed:
Bluetooth: controller: remove hidden GPIOTE configs

The GPIOTE channels used to manage radio front end module GPIOs are
configured with invisible Kconfig symbols. They are therefore not user
visible. Remove them by redefining them in terms of equivalent macros
defined in radio_nrf5.h. This lets us get rid of a couple of ifdefs
that are no longer needed as well.

This is because these invisible symbols depend on the
CONFIG_BT_CTLR_GPIO_* pin configuration symbols. We would like to move
all of that pin configuration to devicetree, because it's hardware
description. It is therefore inconvenient to have the GPIOTE channel
configuration set via Kconfig in a way that depends on something we
would like to remove.

No functional changes or user impact expected: invisible symbols
cannot be set in .conf files.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
Bluetooth: controller: add GPIO polarity indirection

Hide the CONFIG_BT_CTLR_GPIO_..._POL_INV Kconfig macros behind
equivalent ones in the HAL_RADIO_ namespace. This removes any direct
usage of the Kconfig macros from the controller HAL ifdeffery, except
for the usage in radio.c where the actual work is done.

Add "NRF21540" in the macros that are specific to that FEM.

That will be convenient when we are getting the GPIO polarity from
devicetree GPIO flags instead.

No functional changes expected.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
dts: bindings: add generic-fem-two-ctrl-pins

This is a helper binding for radio front-end modules that have a
two-pin control interface, where one pin is used to turn on an power
amplifier (PA) for TX, and another is used to turn on a low-noise
amplifier (LNA) for RX.

Such hardware is already supported by the Bluetooth subsystem, but via
PA/LNA Kconfig settings. Since this is hardware configuration, it is
better to move this to devicetree instead.

Add a binding that makes it possible to define nodes which contain the
same information, along with a bit of extra information related to the
gain in dB of each amplifier not currently available from Kconfig.
This is similar to the existing binding for the nordic,nrf21540-fem
compatible.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>
Signed-off-by: Andrzej Kuros <andrzej.kuros@nordicsemi.no>

mbolivar-nordic committed:
dts: bindings: improve nordic,nrf21540-fem docs

The way the nRF21540 device is configured makes sense once you
understand it, but it's quite unusual: a single peripheral is
configured with two separate devicetree nodes linked by a phandle.

Since this risks entering "exploding head" territory for beginners, it
deserves a thorough example. Add one to the binding's description.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
Bluetooth: controller: remove obsolete Kconfig options

Remove now-unused Kconfig options used for specifying radio front end
module configuration.

These are now handled in devicetree.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
dts: nordic,nrf-radio: add fem property

This will be used to take a phandle to a FEM device, either a generic
two pin device or an nRF21540. Keep the nRF21540 binding example up to
date.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
Bluetooth: controller: support FEMs via devicetree

A FEM in this case is an external Front-End Module. This is basically
a range booster with some extra amplification in the TX and RX paths.

This feature is supported for the Nordic open source controller. See
the changes to the ubx_bmd345eval_nrf52840 board for a quick summary
of what you have to do to take advantage of this.

(This patch doesn't actually remove the feature in the openisa HAL,
though, because that never implemented the required GPIO handling
routines.)

We are using the 'fem' property previously added to the
nordic,nrf-radio's devicetree binding, which points to the FEM that is
in use. If you have this property and the node it points to is
enabled, the controller subsystem respects it. Otherwise, FEM support
is disabled.

This obsoletes the following Kconfig options, which are now unused:

- CONFIG_BT_CTLR_GPIO_PA
- CONFIG_BT_CTLR_GPIO_PA_PIN
- CONFIG_BT_CTLR_GPIO_PA_POL_INV
- CONFIG_BT_CTLR_GPIO_PA_OFFSET

- CONFIG_BT_CTLR_GPIO_LNA
- CONFIG_BT_CTLR_GPIO_LNA_PIN
- CONFIG_BT_CTLR_GPIO_LNA_POL_INV
- CONFIG_BT_CTLR_GPIO_LNA_OFFSET

- CONFIG_BT_CTLR_FEM_NRF21540
- CONFIG_BT_CTLR_GPIO_PDN_PIN
- CONFIG_BT_CTLR_GPIO_PDN_POL_INV
- CONFIG_BT_CTLR_GPIO_CSN_PIN
- CONFIG_BT_CTLR_GPIO_CSN_POL_INV
- CONFIG_BT_CTLR_GPIO_PDN_CSN_OFFSET

The PA and LNA pins are now specified via fem-specific devicetree
properties in the FEM node:

- The "generic" PA/LNA case is handled with the ctx-gpios and
  crx-gpios properties of the generic-fem-two-ctrl-pins compatible,
  respectively.

- If the fem is an nRF21540, use the tx-en-gpios and rx-en-gpios
  properties instead (and also respectively). This allows us to specify
  FEM properties in a way that makes sense for the hardware datasheet,
  while still handling them in a uniform way within the controller.

We support this in the nRF5 HAL with a new radio_nrf5_fem.h
sub-header, which pulls in radio_nrf5_fem_generic.h or
radio_nrf5_fem_nrf21540.h depending on the fem node's compatible, if
one is defined. These in turn let us replace the implementation
routines in radio.c with DT equivalents.

Keep in-tree users and devicetree binding documentation up to date.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
Bluetooth: controller: add FEM offset indirection

Hide the CONFIG_BT_CTLR_GPIO_..._OFFSET Kconfig macros behind
equivalent ones in the HAL_RADIO_ namespace. This removes any direct
usage of the Kconfig macros from the controller HALs.

That will be convenient when we are getting these values from
devicetree properties instead.

Notice that the "PDN/CSN offset" is dropping the "CSN" portion of the
name: CONFIG_BT_CTLR_GPIO_PDN_CSN_OFFSET is now
HAL_RADIO_GPIO_PDN_OFFSET. This is because the DT binding for nRF21540
has a pre-existing 'pdn-settle-time-us' property we're going to use to
replace the Kconfig option.

Other than that, the name changes are self-explanatory.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
Bluetooth: controller: add PA/LNA existence indirection

Add indirection macros for the ifdefs that test for the existence of
PA and LNA pins.

This will be convenient when the pins are configured in devicetree,
and the Kconfig symbols will no longer be around.

No functional changes expected.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
Bluetooth: controller: add nRF21540-as-fem indirection

Introduce an intermediate macro for use detecting that we have a FEM
node, and it's an nRF21540. This is also prep work for converting the
FEM configuration mechanism to devicetree.

No functional changes expected.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio.c:
```diff
@@ -23,25 +26,37 @@
+#define NRF_GPIO_PA	       ((NRF_GPIO_Type*)DT_REG_ADDR(NRF_GPIO_PA_CTLR))
+#define NRF_GPIO_PA_PIN        DT_GPIO_PIN(PA_NODE, pa_gpios)
+#define NRF_GPIO_PA_FLAGS      DT_GPIO_FLAGS(PA_NODE, pa_gpios)
+#define NRF_GPIO_PA_ACTIVE_LOW (NRF_GPIO_PA_FLAGS & GPIO_ACTIVE_LOW)
```
> Could we have some documentation (internal or any other reference) ...

I've improved the commit logs and comments in the latest draft.

> Specific, `NRF_GPIO_PA_CTLR` what is the purpose?

It's a GPIO controller node identifier. See https://docs.zephyrproject.org/latest/reference/devicetree/api.html#gpio. We need it here to distinguish port 0 from port 1.

> Why, `NRF_GPIO_PA_PSEL` and `NRF_GPIO_PA_PIN` both mean a pin index?

They don't. A PSEL value is not a pin number. See the comments here for an explanation of the difference between the two: https://github.com/zephyrproject-rtos/zephyr/blob/1238410914597aef1e0661e0b2e240a448f2dd60/soc/arm/nordic_nrf/soc_nrf_common.h#L95

Part of the point of this PR is to remove the "port and pin in a single integer" encoding for pins in preference of the standard devicetree `<&gpio_controller pin_number flags>;` type of property.

mbolivar-nordic commented:
Sorry for the long delay. I've updated this.

With the following app.overlay in the peripheral_hr sample:

```
$ cat app.overlay 
&arduino_spi {
	fem_spi_if: fem-spi@0 {
		label = "FEM_SPI";
		compatible = "nordic,nrf21540-fem-spi";
                spi-max-frequency = <8000000>;
		reg = <0>;
	};
};

/ {
	nrf21540: fem {
		compatible = "nordic,nrf21540-fem";
		tx-en-gpios = <&gpio1 10 0>;
		rx-en-gpios = <&gpio1 11 0>;
		pdn-gpios = <&gpio0 27 GPIO_ACTIVE_LOW>;
		spi-if = <&fem_spi_if>;
	};
};

&radio {
	fem = <&nrf21540>;
};
```

I get the following behavior on nRF52840-DK, which seems reasonable to me:

![Screenshot from 2021-11-08 22-01-57](https://user-images.githubusercontent.com/62298616/140870939-c76da6d7-75a9-47c8-9b4f-1ef1b782c295.png)

and with:

```
$ cat app.overlay
/ {
	generic: fem {
		compatible = "generic-fem-two-ctrl-pins";
		ctx-gpios = <&gpio1 10 0>;
		ctx-settle-time-us = <5>;
		crx-gpios = <&gpio1 11 0>;
		crx-settle-time-us = <5>;
	};
};

&radio {
	fem = <&generic>;
};
```

I get:

![Screenshot from 2021-11-08 22-05-09](https://user-images.githubusercontent.com/62298616/140871199-46494c83-3b17-4c0c-8a6e-007c07dde59a.png)

I am a bit unsure about this part in radio.c, though:

```c
/* TODO is this OK? */
#define NRF_GPIO_PDN_OFFSET 0
```

If someone (@hubertmis ?) can confirm this is OK, I think this is ready to take out of draft.

hubertmis commented:
> I am a bit unsure about this part in radio.c, though:
> 
> ```c
> /* TODO is this OK? */
> #define NRF_GPIO_PDN_OFFSET 0
> ```
> 
> If someone (@hubertmis ?) can confirm this is OK, I think this is ready to take out of draft.

It should work, but I would rework it a little. Inside `void radio_gpio_pa_lna_enable(uint32_t trx_us)` there is a `#if` to configure hardware associated with PDN pin only when `FEM_IS_NRF21540`. I would change the condition here to something like `#if defined(HAL_RADIO_FEM_IS_NRF21540) && DT_NODE_HAS_PROP(FEM_NODE, pdn_gpios)`. This way you could remove `#define NRF_GPIO_PDN_OFFSET 0` from `#else` block, because its value would be unused if PDN pin is not configured.

mbolivar-nordic commented:
>  would change the condition here

Done thanks.

mbolivar-nordic commented:
> In all FEMs of which I am aware, SoC needs to manage bypass signals. Compiling out software support for FEM if FEM is present in RF path would not work for such devices.

The latest version reflects this comment from @hubertmis .

mbolivar-nordic committed:
doc: release notes: fix format

This file has some formatting issues, maybe due to some errors after
copy/pasting from the previous release. Fix them.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
doc: release notes: BT FEM Kconfig removal

Document what happened to the FEM-related Bluetooth controller Kconfig
options.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic commented:
Rebased and added release notes.

mbolivar-nordic commented:
Addressed a compilation issue on nRF51.

carlescufi commented:
@anangl @hubertmis @cvinayak could we get a review from you please?

carlescufi commented:
@brec-u-blox would you mind taking a look, since the only board affected is yours?

mbolivar-nordic commented:
I'm aware of the merge conflicts and will fix them when we have enough reviewers.

carlescufi reviewed (approved):
ACK from my side, but do not merge until @cvinayak has approved.

anangl reviewed (approved):
LGTM. Only a few nitpicking comments.

brec-u-blox commented:
> 
> 
> @brec-u-blox would you mind taking a look, since the only board affected is yours?

LGTM, though the FEM settle time can be reduced to 1uS (Skyworks RFX2411)
![image](https://user-images.githubusercontent.com/13458841/141216043-1b5a6266-b5a6-4850-a3f4-1d71cedee3d5.png)
Tested OK here on a ubx_bmd345eval_nrf52840.

cvinayak reviewed (approved):
As @brec-u-blox has confirmed on hardware, and I do not have any requests from my end, approved. (if anything comes up in the future, when I test/use the feature or any bugs reported, it can be addressed accordingly after this PR is merged).

cfriedt commented:
@mbolivar-nordic - please rebase and resolve conflicts

mbolivar-nordic committed:
boards: ubx_bmd345eval_nrf52840: shorten FEM settle times

Based on a report from Bob Recny, reducing the settling times doesn't
break anything.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic commented:
> @mbolivar-nordic - please rebase and resolve conflicts

Done.

> LGTM, though the FEM settle time can be reduced to 1uS (Skyworks RFX2411)

@brec-u-blox The board files did not specify CONFIG_BT_CTLR_GPIO_PA_OFFSET or CONFIG_BT_CTLR_GPIO_LNA_OFFSET, so they currently (in main) get the default values of 5 for both.

The patch in this PR which moves FEM config to DT is using 5 to try to avoid a change in behavior in a patch that was only meant to provide a change in configuration mechanism.

Since from your report, it's safe to lower the value, I've done so in a separate commit.

mbolivar-nordic commented:
> LGTM. Only a few nitpicking comments.

Addressed, thanks

brec-u-blox commented:
@mbolivar-nordic Thanks for the clarification and separate commit.
